src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayTranslator.cs

```diff
@@ -54,108 +60,117 @@ public class NpgsqlArrayTranslator : IMethodCallTranslator, IMemberTranslator
-                        typeof(int)),
-                    _sqlExpressionFactory.Constant(0));
-            }
+            // The array/list CLR type may be mapped to a non-array database type (e.g. byte[] to bytea, or just
+            // value converters) - we don't want to translate for those cases.
+            static bool IsMappedToNonArray(SqlExpression arrayOrList)
+                => arrayOrList.TypeMapping is RelationalTypeMapping typeMapping &&
+                   !(typeMapping is NpgsqlArrayTypeMapping || typeMapping is NpgsqlJsonTypeMapping);
 
-            // Note that .Where(e => new[] { "a", "b", "c" }.Any(p => e.SomeText == p)))
-            // is pattern-matched in AllAnyToContainsRewritingExpressionVisitor, which transforms it to
-            // new[] { "a", "b", "c" }.Contains(e.Some Text).
-
-            if (method.IsClosedFormOf(Contains) &&
-                (
-                    // Handle either parameters (no mapping but supported CLR type), or array columns. We specifically
-                    // don't want to translate if the type mapping is bytea (CLR type is array, but not an array in
-                    // the database).
-                    array.TypeMapping == null && _typeMappingSource.FindMapping(array.Type) != null ||
-                    array.TypeMapping is NpgsqlArrayTypeMapping
-                ) &&
-                // Exclude arrays/lists over Nullable<T> since the ADO layer doesn't handle them (but will in 5.0)
-                Nullable.GetUnderlyingType(elementType) == null)
+            SqlExpression TranslateCommon(SqlExpression arrayOrList, IReadOnlyList<SqlExpression> arguments)
             {
-                var item = arguments[1];
+                // Predicate-less Any - translate to a simple length check.
+                if (method.IsClosedFormOf(EnumerableAnyWithoutPredicate))
+                {
+                    return _sqlExpressionFactory.GreaterThan(
+                        _jsonPocoTranslator.TranslateArrayLength(arrayOrList) ??
+                        _sqlExpressionFactory.Function(
+                            "cardinality",
+                            new[] { arrayOrList },
+                            nullable: true,
+                            argumentsPropagateNullability: TrueArrays[1],
+                            typeof(int)),
+                        _sqlExpressionFactory.Constant(0));
+                }
 
-                switch (array)
+                // Note that .Where(e => new[] { "a", "b", "c" }.Any(p => e.SomeText == p)))
+                // is pattern-matched in AllAnyToContainsRewritingExpressionVisitor, which transforms it to
+                // new[] { "a", "b", "c" }.Contains(e.Some Text).
+
+                if ((method.IsClosedFormOf(EnumerableContains) || // Enumerable.Contains extension method
+                     method.Name == nameof(List<int>.Contains) && method.DeclaringType.IsGenericList() &&
```

Comment:
Should `DeclaringType` be assignable to `IReadOnlyCollection<T>` or `ICollection<T>` which own that member?

Reply:
Unless I'm mistaken, if the user does Contains over an ICollection or similar, they get the Enumerable.Contains extension method (which is checked for just above). The second line checks specifically for List<T>.Contains, does that make sense?

Reply:
Extension methods have less priority than members.

Reply:
Of course, but I'm not following what you're saying. If the CLR type is List<T>, then we get List. Contains (the second line above). Otherwise, for any other IEnumerable<T>, we get Enumerable.Contains (the first line). Where's the problem?

Reply:
If someone uses a read only collection which in public API, but under the hood it's an array or a list.

Reply:
Why is that a problem?

Reply:
My bad, thought that read only collections have `Contains` as a member, but `IReadOnlyDictionary<,>` has such members. Not a problem if a property isn't `ICollection<T>`:

```csharp
public class Entity
{
    private List<int> _field;
    public ICollection<int> Property => _field;
}
```
```csharp
context.Entitites
    .Where(e => e.Property.Contains(value)); // Contains is a member function
```